### ISSUE TITLE
`bounding_box.clip_to_image()` should set class_id to -1 for boxes entirely out of the image

### ISSUE THREAD

**Issue by LukeWood** [keras-team/keras-cv#801]
Basically, Shear/Mosaic can clip bounding boxes entirely outside of the bounds of an image.  clip_to_image() should set the class id of the bounding boxes to -1, instead of having an empty box.  Right now, these are clipped to be at the edges of the image with a width/height of 0 - but this is nonsense and causes NaNs in the RetinaNet.

https://github.com/keras-team/keras-cv/pull/799/files#diff-9178d5b7485015d2c0420576c0b5998cb3f2de4e1e5bb8bef3fa3371daab7614R250

**Comment by bhack** [comment]
Do you have any reference that using "a garbage class" is a good solution and it will not create performance issues to the model?

**Comment by bhack** [comment]
> and causes NaNs in the RetinaNet.

I cannot find this "practice garbage class" practice in https://github.com/tensorflow/models/blob/master/official/vision/ops/preprocess_ops.py

https://github.com/tensorflow/models/blob/master/official/vision/ops/preprocess_ops.py
https://github.com/tensorflow/models/blob/master/official/vision/modeling/retinanet_model.py
https://github.com/tensorflow/models/blob/master/official/vision/tasks/retinanet.py

How they are going to reproduce reference results?

**Comment by LukeWood** [comment]
> Do you have any reference that using "a garbage class" is a good solution and it will not create performance issues to the model?

Im not sure what you mean here.  -1 indicates no box when the number of boxes between images does not match in a batched bounding box tensor.  This is true across almost all OD pipelines.  Otherwise you have to use ragged tensors everywhere - which don’t work on TPU

**Comment by bhack** [comment]
What I meant is this "garbage" class going in the loss?

**Comment by LukeWood** [comment]
> What I meant is this "garbage" class going in the loss?

Already filtered out with a tf.where()

this is how background classes are already encoded
